### ISSUE TITLE
NMS is not working

### ISSUE THREAD

**Issue by twinanda** [fizyr/keras-retinanet#680]
I am having issue with the NMS. My network trains just fine, but when I use the network to predict, I get a lot of spurious boxes on the object. It seems that the NMS is not working. I am using the version from 3 days ago. Anybody has the same issue or has a solution to this?

**Comment by hgaiser** [comment]
Could you show an image of your problem? Are the classes the same for the overlapping objects ?

**Comment by twinanda** [comment]
![sketch 8](https://user-images.githubusercontent.com/17176115/45550667-d2877f80-b82b-11e8-8596-540eb603f002.png)
Trying to detect a surgical tool. Only one class in the project. I am showing all bounding boxed with >0.35 confidence.

**Comment by twinanda** [comment]
@hgaiser any comments on this?

**Comment by twinanda** [comment]
@hgaiser Yes this is a custom code and I am using retinanet implementation in the backend. I created the models using create_models function in the package which returns the cpu model, gpu model and inference model.

backbone = models.backbone("resnet50")
weights = backbone.download_imagenet()
model, training_model, prediction_model = create_models(backbone_retinanet=backbone.retinanet, num_classes=self.num_classes, weights=weights, multi_gpu=1, freeze_backbone=False)

And the rest of the code is just typical Keras training and inference in which I didn’t change anything in the model.

**Comment by hgaiser** [comment]
Could you try running with `evaluate.py`, with the `--save-path` (and probably `--score-threshold`) flag enabled?

**Comment by twinanda** [comment]
@hgaiser I am not sure what `evaluate.py` takes as input arguments. 
Could you help me with the terminal command to run the evaluate.py with my own model? Assuming my model is located in `$model_path` and my image to test is `$image_path`, how do I execute this with `evaluate.py` on my terminal?

**Comment by peacherwu** [comment]
> @hgaiser I am not sure what `evaluate.py` takes as input arguments.
> Could you help me with the terminal command to run the evaluate.py with my own model? Assuming my model is located in `$model_path` and my image to test is `$image_path`, how do I execute this with `evaluate.py` on my terminal?

Try:
`/path/to/evaluate.py -h`

**Comment by peacherwu** [comment]
> @hgaiser I am not sure what `evaluate.py` takes as input arguments.
> Could you help me with the terminal command to run the evaluate.py with my own model? Assuming my model is located in `$model_path` and my image to test is `$image_path`, how do I execute this with `evaluate.py` on my terminal?

Look into
`/path/to/evaluate.py -h`

**Comment by twinanda** [comment]
@hgaiser I saw that in draw_detection function, the default value for score_threshold is 0.5. Even though, the default parameter in other function score_threshold is 0.05. When I changed draw_detection in _get_detections from this:
`draw_detections(raw_image, image_boxes, image_scores, image_labels, label_to_name=generator.label_to_name)`
to 
`draw_detections(raw_image, image_boxes, image_scores, image_labels, label_to_name=generator.label_to_name, score_threshold=score_threshold)`
I could see that the NMS in the network does not really work. This the image that is drawn:
![12](https://user-images.githubusercontent.com/17176115/46001304-40913980-c0ab-11e8-9e3f-d87fb6b03bb4.png)

CMIIW, but if the NMS works, the network should yield fewer bounding boxes than these.

**Comment by hgaiser** [comment]
Sorry, been pretty busy :)

Yeah, those boxes should be filtered for sure. The command I normally run for evaluation is something along the lines of : 
```shell
python keras_retinanet/bin/evaluate.py --score-threshold 0.5 --save-path output csv /path/to/train.csv /path/to/classes.csv
```
The images should then be saved to `output` directory (might need to create it before running this command). You need a dataset to run the evaluate command, you can use your training dataset if you don't have a test set, it's just to check if NMS is doing its thing. 

Also, just to be sure, what TF and Keras are you using ? How are you creating the model? Just by using `train.py`?

Did you in the end change the score_threshold to 0.5 for the image you shared? Because the detections seem to have scores < 0.5 too.

**Comment by twinanda** [comment]
@hgaiser No worries, I figured you were. It was just a friendly reminder. Thank you for taking the time to look into this. :)

I got the evaluate.py to work, that image I posted was generated by the evaluate script. But thanks for the instructions. I did exactly just what you explained.

My network was trained on Keras 2.1.6 and TF 1.8.0. But, your script requires 2.2.0, so I just upgraded Keras to 2.2.0 without changing TF. I am creating the model by calling create_models function in train.py. Train the training model using fit_generator function. And perform inference using the inference model. 

So the image that I showed you is generated using 0.05 score_threshold. With 0.5 confidence threshold, there is no detection past that threshold. 

![](https://user-images.githubusercontent.com/17176115/46001304-40913980-c0ab-11e8-9e3f-d87fb6b03bb4.png)

In any case, shouldn't these detections be removed by the nms function anyway, even though their scores are below 0.5? 
FYI, in my application, my confidence threshold is lower than 0.5 to boost recall, but these are the detections given by the network. 

![](https://user-images.githubusercontent.com/17176115/45550667-d2877f80-b82b-11e8-8596-540eb603f002.png)

**Comment by fernandocamargoti** [comment]
I'm having the same issue:

![103](https://user-images.githubusercontent.com/3669312/46154313-49207600-c24b-11e8-9e9c-41ad2ac12410.png)

Even with 0.5 of threshold for the score, I often find one bounding box completly inside another one.

I'm creating the model as follows:

`model = models.backbone("resnet50").retinanet(num_classes=1)`

And:

`retinanet_bbox(model)`

**Comment by hgaiser** [comment]
Hmm okay I'll try to look into this a bit further. Can either (or both?) of you host your trained model somewhere and an image I can test on? Also list the image-min/image-max sizes used during training please :)

**Comment by fernandocamargoti** [comment]
https://drive.google.com/open?id=1-N0FRGVX8gDLx9LQTR1xAQM6k8Cn7sO4

I used all images with 512x512px during training (the same size of this sample image).

**Comment by hgaiser** [comment]
> https://drive.google.com/open?id=1-N0FRGVX8gDLx9LQTR1xAQM6k8Cn7sO4
> 
> I used all images with 512x512px during training (the same size of this sample image).

Hmm your model is a bit weird, it produces insanely large regression values. 

Here are some regression outputs for your model:
```
[[[ -8.302955  -37.769913  -29.88107     4.029128 ]
  [ -3.35301   -27.618559  -28.103317    7.8425407]
  [  2.4992073 -15.7645855 -20.98438    17.89835  ]
  ...
  [ -4.050961    3.703164   11.332778   -3.4172819]
  [ -3.7851362   1.8371779   7.196213   -0.9980116]
  [ -6.2028513  -1.6195434   6.1281867  -1.8843968]]]
```

Compared to the values of some COCO model I had laying around:

```
[[[ 0.30758613  0.16628824 -0.03616999  0.00807128]
  [ 0.23593889  0.14860313  0.06976658 -0.05330962]
  [ 0.12915191  0.09145881  0.13056836 -0.18028024]
  ...
  [ 0.00088133 -0.22867756 -0.0994329  -0.17714852]
  [-0.03921628 -0.2937888  -0.13039753 -0.11414814]
  [ 0.05310528 -0.22824721 -0.09401623 -0.05218803]]]
```

These are more or less factors of the width and height to shift the anchor coordinates with, those large values are really weird. I expect it to be below 1.0.

What image preprocessing did you use? I was using the jupyter notebook to test your network.

For the record, what was your `tensorflow` version?

**Comment by fernandocamargoti** [comment]
@hgaiser, I'm using the "caffe" preprocessing present in your library and Tensorflow 1.9.0.

Could it mean that the default anchors are not well suited for this problem?

**Comment by fernandocamargoti** [comment]
@hgaiser 
I understand that anchors are not related with preprocessing. What I meant was when you said that the regression outputs were weird. I wonder what would be the cause and if it could be related to the anchors.

**Comment by void32** [comment]
I see this problem as well:
![nms_isue_clip](https://user-images.githubusercontent.com/6043260/46470493-1809ed80-c7d7-11e8-9dd3-59c841ff9aef.png)

Is there are a recommended way to edit the nms and score threshold. Currently I'm changing their value directly in filter_detections.py but is there a config file for this?

I'm using tensorflow 1.6 and Keras 2.2.3 btw.

**Comment by hgaiser** [comment]
@void32 a config for these parameters is still missing.

Could you try updating tensorflow to 1.8 or higher? I've seen some issues when using older tensorflow versions.

**Comment by void32** [comment]
Ok, but it looks like some work already has been done to implement a config:
`[filter_detections.py#L207](https://github.com/fizyr/keras-retinanet/blob/078e12d87ecc9422f525965c0dc1a9478117c480/keras_retinanet/layers/filter_detections.py#L207)`
Is there a preferred path for the config ini-file?
 
I tried it with tensorflow 1.11 and got same behavior. However I calculated the IoU for the overlapping boxes and turn out to be below 0.50 (which is the threshold in [filter_detections.py#L29](https://github.com/fizyr/keras-retinanet/blob/078e12d87ecc9422f525965c0dc1a9478117c480/keras_retinanet/layers/filter_detections.py#L29)):
![ious](https://user-images.githubusercontent.com/6043260/46731724-bf7d9900-cc8b-11e8-83a1-c03fc91e0ac3.png)

**Comment by hgaiser** [comment]
Did you calculate the IOU before or after correcting for the image shape? If after, could you try computing then before the correction? That's also what the network does.

**Comment by void32** [comment]
I used the boxes from the model output to calculate intersection and union area to get the IoU. The correction, is that the resizing of the image?

**Comment by hgaiser** [comment]
Yeah the image is resized, so the boxes are computed within the resized image. The example script scales the boxes back so that they match the original image shape.

**Comment by void32** [comment]
In our setup, unlike in [ResNet50RetinaNet.ipynb](https://github.com/fizyr/keras-retinanet/blob/master/examples/ResNet50RetinaNet.ipynb) I don't resize the images explicitly. However, test and training images are vastly from the same camera and stored with the same dimensions.
Anyways, if I don't resize the images I guess there is no correction and IoU should be right ?

**Comment by hgaiser** [comment]
But do you use our generators? They resize the images by default.

**Comment by void32** [comment]
For training, I use the  _keras-retinanet/keras_retinanet/bin/train.py_ script and for prediction i use the predict_on_batch method in _models_. I'm not sure if those use generators?

**Comment by hgaiser** [comment]
Yeah they use generators. By default they resize your images such that the smallest side is 800px long. If that means that the larger side is then larger than 1333, it will downsize such that this side is exactly 1333px.

If you didn't do anything other than train, the example shows how to run inference on an image and visualize the results, taking into account the resizing of the image: https://github.com/fizyr/keras-retinanet/blob/master/examples/ResNet50RetinaNet.ipynb

**Comment by hgaiser** [comment]
So any update on this? NMS seems to work fine in my cases.

**Comment by void32** [comment]
Sorry for leaving this thread hanging, had to help out on another project.

Anyways the NMS works for me as well, also when I compute the on IoU for not scaled boxes.
![nms_test](https://user-images.githubusercontent.com/6043260/47282372-406b5780-d5df-11e8-837b-ad720746a350.png)

**Comment by hgaiser** [comment]
Alright, then I'll close the issue since it appears to work. Feel free to reopen if there is something to be added or if you feel I'm wrong :)

**Comment by michaelt88** [comment]
Hi I am having a similar issue where I have many overlapping boxes.
I change the nms_threshold value in filter_detections.py, and re-run the conversion to generate an inference model, but I don't see any reduction in overlapping boxes.
I couldn't quite understand from reading this thread how this was resolved, was it to do with how the images are being resized? I am feeding in 1000x1000px images.
![multiple detections](https://user-images.githubusercontent.com/36904934/50258431-a71fbc80-043a-11e9-9357-c9ddac5398c4.jpeg)

**Comment by michaelt88** [comment]
If I increase the detection threshold to 0.4, then I get rid of a lot of the overlapping boxes, but several of the targets are missed
![increased detection score threshold](https://user-images.githubusercontent.com/36904934/50262761-cf64e680-044d-11e9-8ec2-2c04deb8487d.jpeg)

**Comment by UpCoder** [comment]
@michaelt88  @hgaiser, Do you solve the problem? I also meet the problem. The score of the detected object is very low. If I increase the score_threshold, many targets will be missed. However, if I do not increase the score_threshold, there are many overlaps.

**Comment by hgaiser** [comment]
The github issues page is not for usage questions. Please use the keras-retinanet Slack channel (see README.md) for usage questions.

Having said that, a low score usually indicates an issue with the training data. Try debug.py and see if the dataset is seemingly correct.
